File: templates/components/workflows/typescript/financial_report/workflow.ts
import { toAgentRunEvent, toSourceEvent } from "@llamaindex/server";
import {
  callTools,
  chatWithTools,
  documentGenerator,
  interpreter,
} from "@llamaindex/tools";
import {
  AgentInputData,
  AgentWorkflowContext,
  BaseToolWithCall,
  ChatMemoryBuffer,
  ChatMessage,
  ChatResponseChunk,
  HandlerContext,
  Metadata,
  NodeWithScore,
  Settings,
  StartEvent,
  StopEvent,
  ToolCall,
  ToolCallLLM,
  Workflow,
  WorkflowEvent,
} from "llamaindex";
import { getIndex } from "../data";

const TIMEOUT = 360 * 1000;
const OUT_DIR = "output/tools";
const UPLOADED_FILE_DIR = "output/uploaded";
const FILE_SERVER_URL_PREFIX = "/api/files";

export async function workflowFactory() {
  const index = await getIndex();

  const queryEngineTool = index.queryTool({
    metadata: {
      name: "query_document",
      description: `This tool can retrieve information about Apple and Tesla financial data`,
    },
    includeSourceNodes: true,
  });

  if (!process.env.E2B_API_KEY) {
    throw new Error("E2B_API_KEY is required to use the code interpreter tool");
  }

  const codeInterpreterTool = interpreter({
    apiKey: process.env.E2B_API_KEY!,
    uploadedFilesDir: UPLOADED_FILE_DIR,
    outputDir: OUT_DIR,
    fileServerURLPrefix: FILE_SERVER_URL_PREFIX,
  });
  const documentGeneratorTool = documentGenerator({
    outputDir: "output/tools",
    fileServerURLPrefix: FILE_SERVER_URL_PREFIX,
  });

  return new FinancialReportWorkflow({
    queryEngineTool,
    codeInterpreterTool,
    documentGeneratorTool,
    timeout: TIMEOUT,
  });
}

// Create a custom event type
class InputEvent extends WorkflowEvent<{ input: ChatMessage[] }> {}

class ResearchEvent extends WorkflowEvent<{
  toolCalls: ToolCall[];
}> {}

class AnalyzeEvent extends WorkflowEvent<{
  input: ChatMessage | ToolCall[];
}> {}

class ReportGenerationEvent extends WorkflowEvent<{
  toolCalls: ToolCall[];
}> {}

const DEFAULT_SYSTEM_PROMPT = `
You are a financial analyst who are given a set of tools to help you.
It's good to using appropriate tools for the user request and always use the information from the tools, don't make up anything yourself.
For the query engine tool, you should break down the user request into a list of queries and call the tool with the queries.
`;

class FinancialReportWorkflow extends Workflow<
  AgentWorkflowContext,
  AgentInputData,
  string
> {
  llm: ToolCallLLM;
  memory: ChatMemoryBuffer;
  queryEngineTool: BaseToolWithCall;
  codeInterpreterTool: BaseToolWithCall;
  documentGeneratorTool: BaseToolWithCall;
  systemPrompt?: string;

  constructor(options: {
    queryEngineTool: BaseToolWithCall;
    codeInterpreterTool: BaseToolWithCall;
    documentGeneratorTool: BaseToolWithCall;
    systemPrompt?: string;
    verbose?: boolean;
    timeout?: number;
  }) {
    super({
      verbose: options?.verbose ?? false,
      timeout: options?.timeout ?? 360,
    });

    this.llm = Settings.llm as ToolCallLLM;
    if (!this.llm.supportToolCall) {
      throw new Error("LLM is not a ToolCallLLM");
    }
    this.systemPrompt = options.systemPrompt ?? DEFAULT_SYSTEM_PROMPT;
    this.queryEngineTool = options.queryEngineTool;
    this.codeInterpreterTool = options.codeInterpreterTool;

    this.documentGeneratorTool = options.documentGeneratorTool;
    this.memory = new ChatMemoryBuffer({ llm: this.llm, chatHistory: [] });

    // Add steps
    this.addStep(
      {
        inputs: [StartEvent<AgentInputData>],
        outputs: [InputEvent],
      },
      this.prepareChatHistory,
    );

    this.addStep(
      {
        inputs: [InputEvent],
        outputs: [
          InputEvent,
          ResearchEvent,
          AnalyzeEvent,
          ReportGenerationEvent,
          StopEvent,
        ],
      },
      this.handleLLMInput,
    );

    this.addStep(
      {
        inputs: [ResearchEvent],
        outputs: [AnalyzeEvent],
      },
      this.handleResearch,
    );

    this.addStep(
      {
        inputs: [AnalyzeEvent],
        outputs: [InputEvent],
      },
      this.handleAnalyze,
    );

    this.addStep(
      {
        inputs: [ReportGenerationEvent],
        outputs: [InputEvent],
      },
      this.handleReportGeneration,
    );
  }

  prepareChatHistory = async (
    ctx: HandlerContext<AgentWorkflowContext>,
    ev: StartEvent<AgentInputData>,
  ): Promise<InputEvent> => {
    const { userInput, chatHistory = [] } = ev.data;
    if (!userInput) throw new Error("Invalid input");

    this.memory.set(chatHistory);

    if (this.systemPrompt) {
      this.memory.put({ role: "system", content: this.systemPrompt });
    }

    this.memory.put({ role: "user", content: userInput });

    const messages = await this.memory.getMessages();
    return new InputEvent({ input: messages });
  };

  handleLLMInput = async (
    ctx: HandlerContext<AgentWorkflowContext>,
    ev: InputEvent,
  ): Promise<
    | InputEvent
    | ResearchEvent
    | AnalyzeEvent
    | ReportGenerationEvent
    | StopEvent<AsyncGenerator<ChatResponseChunk, any, any> | undefined>
  > => {
    const chatHistory = ev.data.input;

    const tools = [
      this.codeInterpreterTool,
      this.documentGeneratorTool,
      this.queryEngineTool,
    ];

    const toolCallResponse = await chatWithTools(this.llm, tools, chatHistory);

    if (!toolCallResponse.hasToolCall()) {
      return new StopEvent(toolCallResponse.responseGenerator);
    }

    if (toolCallResponse.hasMultipleTools()) {
      this.memory.put({
        role: "assistant",
        content:
          "Calling different tools is not allowed. Please only use multiple calls of the same tool.",
      });
      const chatHistory = await this.memory.getMessages();
      return new InputEvent({ input: chatHistory });
    }

    // Put the LLM tool call message into the memory
    // And trigger the next step according to the tool call
    if (toolCallResponse.toolCallMessage) {
      this.memory.put(toolCallResponse.toolCallMessage);
    }
    const toolName = toolCallResponse.getToolNames()[0];
    switch (toolName) {
      case this.codeInterpreterTool.metadata.name:
        return new AnalyzeEvent({
          input: toolCallResponse.toolCalls,
        });
      case this.documentGeneratorTool.metadata.name:
        return new ReportGenerationEvent({
          toolCalls: toolCallResponse.toolCalls,
        });
      default:
        if (this.queryEngineTool.metadata.name === toolName) {
          return new ResearchEvent({
            toolCalls: toolCallResponse.toolCalls,
          });
        }
        throw new Error(`Unknown tool: ${toolName}`);
    }
  };

  handleResearch = async (
    ctx: HandlerContext<AgentWorkflowContext>,
    ev: ResearchEvent,
  ): Promise<AnalyzeEvent> => {
    ctx.sendEvent(
      toAgentRunEvent({
        agent: "Researcher",
        text: "Researching data",
        type: "text",
      }),
    );

    const { toolCalls } = ev.data;

    const toolMsgs = await callTools({
      tools: [this.queryEngineTool],
      toolCalls,
      writeEvent: (text, step) => {
        ctx.sendEvent(
          toAgentRunEvent({
            agent: "Researcher",
            text,
            type: toolCalls.length > 1 ? "progress" : "text",
            current: step,
            total: toolCalls.length,
          }),
        );
      },
    });
    for (const toolMsg of toolMsgs) {
      this.memory.put(toolMsg);
    }

    const sourcesNodes: NodeWithScore<Metadata>[] = toolMsgs
      .map((msg) => (msg.options as any)?.toolResult?.result?.sourceNodes)
      .flat()
      .filter(Boolean);

    if (sourcesNodes.length > 0) {
      ctx.sendEvent(toSourceEvent(sourcesNodes));
    }

    return new AnalyzeEvent({
      input: {
        role: "assistant",
        content:
          "I have finished researching the data, please analyze the data.",
      },
    });
  };

  /**
   * Analyze a research result or a tool call for code interpreter from the LLM
   */
  handleAnalyze = async (
    ctx: HandlerContext<AgentWorkflowContext>,
    ev: AnalyzeEvent,
  ): Promise<InputEvent> => {
    ctx.sendEvent(
      toAgentRunEvent({
        agent: "Analyst",
        text: "Analyzing data",
        type: "text",
      }),
    );
    // Request by workflow LLM, input is a list of tool calls
    let toolCalls: ToolCall[] = [];
    if (Array.isArray(ev.data.input)) {
      toolCalls = ev.data.input;
    } else {
      // Requested by Researcher, input is a ChatMessage
      // We start new LLM chat specifically for analyzing the data
      const analysisPrompt = `
      You are an expert in analyzing financial data.
      You are given a set of financial data to analyze. Your task is to analyze the financial data and return a report.
      Your response should include a detailed analysis of the financial data, including any trends, patterns, or insights that you find.
      Construct the analysis in textual format; including tables would be great!
      Don't need to synthesize the data, just analyze and provide your findings.
      `;

      // Clone the current chat history
      // Add the analysis system prompt and the message from the researcher
      const chatHistory = await this.memory.getMessages();
      const newChatHistory = [
        ...chatHistory,
        { role: "system", content: analysisPrompt },
        ev.data.input,
      ];
      const toolCallResponse = await chatWithTools(
        this.llm,
        [this.codeInterpreterTool],
        newChatHistory as ChatMessage[],
      );

      if (!toolCallResponse.hasToolCall()) {
        this.memory.put(await toolCallResponse.asFullResponse());
        const chatHistory = await this.memory.getMessages();
        return new InputEvent({ input: chatHistory });
      } else {
        this.memory.put(toolCallResponse.toolCallMessage!);
        toolCalls = toolCallResponse.toolCalls;
      }
    }

    // Call the tools
    const toolMsgs = await callTools({
      tools: [this.codeInterpreterTool],
      toolCalls,
      writeEvent: (text, step) => {
        ctx.sendEvent(
          toAgentRunEvent({
            agent: "Analyst",
            text,
            type: toolCalls.length > 1 ? "progress" : "text",
            current: step,
            total: toolCalls.length,
          }),
        );
      },
    });
    for (const toolMsg of toolMsgs) {
      this.memory.put(toolMsg);
    }

    const chatHistory = await this.memory.getMessages();
    return new InputEvent({ input: chatHistory });
  };

  handleReportGeneration = async (
    ctx: HandlerContext<AgentWorkflowContext>,
    ev: ReportGenerationEvent,
  ): Promise<InputEvent> => {
    const { toolCalls } = ev.data;

    const toolMsgs = await callTools({
      tools: [this.documentGeneratorTool],
      toolCalls,
      writeEvent: (text, step) => {
        ctx.sendEvent(
          toAgentRunEvent({
            agent: "Reporter",
            text,
            type: toolCalls.length > 1 ? "progress" : "text",
            current: step,
            total: toolCalls.length,
          }),
        );
      },
    });
    for (const toolMsg of toolMsgs) {
      this.memory.put(toolMsg);
    }
    const chatHistory = await this.memory.getMessages();
    return new InputEvent({ input: chatHistory });
  };
}

File: packages/create-llama/templates/components/use-cases/python/financial_report/workflow.py
import os
from typing import List, Optional

from app.index import get_index
from llama_index.core import Settings
from llama_index.core.base.llms.types import ChatMessage, MessageRole
from llama_index.core.llms.function_calling import FunctionCallingLLM
from llama_index.core.memory import ChatMemoryBuffer
from llama_index.core.tools import FunctionTool, QueryEngineTool, ToolSelection
from llama_index.core.workflow import (
    Context,
    Event,
    StartEvent,
    StopEvent,
    Workflow,
    step,
)
from llama_index.server.api.models import AgentRunEvent, ChatRequest
from llama_index.server.settings import server_settings
from llama_index.server.tools.document_generator import DocumentGenerator
from llama_index.server.tools.index import get_query_engine_tool
from llama_index.server.tools.interpreter import E2BCodeInterpreter
from llama_index.server.utils.agent_tool import (
    call_tools,
    chat_with_tools,
)


def create_workflow(chat_request: Optional[ChatRequest] = None) -> Workflow:
    index = get_index(chat_request=chat_request)
    if index is None:
        raise ValueError(
            "Index is not found. Try run generation script to create the index first."
        )
    query_engine_tool = get_query_engine_tool(index=index)
    e2b_api_key = os.getenv("E2B_API_KEY")
    if e2b_api_key is None:
        raise ValueError(
            "E2B_API_KEY is required to use the code interpreter tool. Please check README.md to know how to get the key."
        )
    code_interpreter_tool = E2BCodeInterpreter(api_key=e2b_api_key).to_tool()
    document_generator_tool = DocumentGenerator(
        file_server_url_prefix=server_settings.file_server_url_prefix,
    ).to_tool()

    return FinancialReportWorkflow(
        query_engine_tool=query_engine_tool,
        code_interpreter_tool=code_interpreter_tool,
        document_generator_tool=document_generator_tool,
        timeout=180,
    )


class InputEvent(Event):
    input: List[ChatMessage]
    response: bool = False


class ResearchEvent(Event):
    input: list[ToolSelection]


class AnalyzeEvent(Event):
    input: list[ToolSelection] | ChatMessage


class ReportEvent(Event):
    input: list[ToolSelection]


class FinancialReportWorkflow(Workflow):
    """
    A workflow to generate a financial report using indexed documents.

    Requirements:
    - Indexed documents containing financial data and a query engine tool to search them
    - A code interpreter tool to analyze data and generate reports
    - A document generator tool to create report files

    Steps:
    1. LLM Input: The LLM determines the next step based on function calling.
       For example, if the model requests the query engine tool, it returns a ResearchEvent;
       if it requests document generation, it returns a ReportEvent.
    2. Research: Uses the query engine to find relevant chunks from indexed documents.
       After gathering information, it requests analysis (step 3).
    3. Analyze: Uses a custom prompt to analyze research results and can call the code
       interpreter tool for visualization or calculation. Returns results to the LLM.
    4. Report: Uses the document generator tool to create a report. Returns results to the LLM.
    """

    _default_system_prompt = """
    You are a financial analyst who are given a set of tools to help you.
    It's good to using appropriate tools for the user request and always use the information from the tools, don't make up anything yourself.
    For the query engine tool, you should break down the user request into a list of queries and call the tool with the queries.
    """
    stream: bool = True

    def __init__(
        self,
        query_engine_tool: QueryEngineTool,
        code_interpreter_tool: FunctionTool,
        document_generator_tool: FunctionTool,
        llm: Optional[FunctionCallingLLM] = None,
        system_prompt: Optional[str] = None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.system_prompt = system_prompt or self._default_system_prompt
        self.query_engine_tool = query_engine_tool
        self.code_interpreter_tool = code_interpreter_tool
        self.document_generator_tool = document_generator_tool
        assert query_engine_tool is not None, (
            "Query engine tool is not found. Try run generation script or upload a document file first."
        )
        assert code_interpreter_tool is not None, "Code interpreter tool is required"
        assert document_generator_tool is not None, (
            "Document generator tool is required"
        )
        self.tools = [
            self.query_engine_tool,
            self.code_interpreter_tool,
            self.document_generator_tool,
        ]
        self.llm: FunctionCallingLLM = llm or Settings.llm  # type: ignore
        assert isinstance(self.llm, FunctionCallingLLM)
        self.memory = ChatMemoryBuffer.from_defaults(llm=self.llm)

    @step()
    async def prepare_chat_history(self, ctx: Context, ev: StartEvent) -> InputEvent:
        self.stream = ev.get("stream", True)
        user_msg = ev.get("user_msg")
        chat_history = ev.get("chat_history")

        if chat_history is not None:
            self.memory.put_messages(chat_history)

        # Add user message to memory
        self.memory.put(ChatMessage(role=MessageRole.USER, content=user_msg))

        if self.system_prompt:
            system_msg = ChatMessage(
                role=MessageRole.SYSTEM, content=self.system_prompt
            )
            self.memory.put(system_msg)

        return InputEvent(input=self.memory.get())

    @step()
    async def handle_llm_input(  # type: ignore
        self,
        ctx: Context,
        ev: InputEvent,
    ) -> ResearchEvent | AnalyzeEvent | ReportEvent | StopEvent:
        """
        Handle an LLM input and decide the next step.
        """
        # Always use the latest chat history from the input
        chat_history: list[ChatMessage] = ev.input
        # Get tool calls
        response = await chat_with_tools(
            self.llm,
            self.tools,  # type: ignore
            chat_history,
        )
        if not response.has_tool_calls():
            if self.stream:
                return StopEvent(result=response.generator)
            else:
                return StopEvent(result=await response.full_response())
        # calling different tools at the same time is not supported at the moment
        # add an error message to tell the AI to process step by step
        if response.is_calling_different_tools():
            self.memory.put(
                ChatMessage(
                    role=MessageRole.ASSISTANT,
                    content="Cannot call different tools at the same time. Try calling one tool at a time.",
                )
            )
            return InputEvent(input=self.memory.get())  # type: ignore
        self.memory.put(response.tool_call_message)
        match response.tool_name():
            case self.code_interpreter_tool.metadata.name:
                return AnalyzeEvent(input=response.tool_calls)
            case self.document_generator_tool.metadata.name:
                return ReportEvent(input=response.tool_calls)
            case self.query_engine_tool.metadata.name:
                return ResearchEvent(input=response.tool_calls)
            case _:
                raise ValueError(f"Unknown tool: {response.tool_name()}")

    @step()
    async def research(self, ctx: Context, ev: ResearchEvent) -> AnalyzeEvent:
        """
        Do a research to gather information for the user's request.
        A researcher should have these tools: query engine, search engine, etc.
        """
        ctx.write_event_to_stream(
            AgentRunEvent(
                name="Researcher",
                msg="Starting research",
            )
        )
        tool_calls = ev.input

        tool_call_outputs = await call_tools(
            ctx=ctx,
            agent_name="Researcher",
            tools=[self.query_engine_tool],
            tool_calls=tool_calls,
        )
        for tool_call_output in tool_call_outputs:
            self.memory.put(
                ChatMessage(
                    role=MessageRole.TOOL,
                    content=tool_call_output.tool_output.content,
                    additional_kwargs={
                        "name": tool_call_output.tool_output.tool_name,
                        "tool_call_id": tool_call_output.tool_call_id,
                    },
                )
            )
        return AnalyzeEvent(
            input=ChatMessage(
                role=MessageRole.ASSISTANT,
                content="Researcher: I've finished the research. Please analyze the result.",
            ),
        )

    @step()
    async def analyze(self, ctx: Context, ev: AnalyzeEvent) -> InputEvent:
        """
        Analyze the research result.
        """
        ctx.write_event_to_stream(
            AgentRunEvent(
                name="Analyst",
                msg="Starting analysis",
            )
        )
        event_requested_by_workflow_llm = isinstance(ev.input, list)
        # Requested by the workflow LLM Input step, it's a tool call
        if event_requested_by_workflow_llm:
            # Set the tool calls
            tool_calls = ev.input
        else:
            # Otherwise, it's triggered by the research step
            # Use a custom prompt and independent memory for the analyst agent
            analysis_prompt = """
            You are a financial analyst, you are given a research result and a set of tools to help you.
            Always use the given information, don't make up anything yourself. If there is not enough information, you can asking for more information.
            If you have enough numerical information, it's good to include some charts/visualizations to the report so you can use the code interpreter tool to generate a report.
            """
            # This is handled by analyst agent
            # Clone the shared memory to avoid conflicting with the workflow.
            chat_history = self.memory.get()
            chat_history.append(
                ChatMessage(role=MessageRole.SYSTEM, content=analysis_prompt)
            )
            chat_history.append(ev.input)  # type: ignore
            # Check if the analyst agent needs to call tools
            response = await chat_with_tools(
                self.llm,
                [self.code_interpreter_tool],
                chat_history,
            )
            if not response.has_tool_calls():
                # If no tool call, fallback analyst message to the workflow
                msg_content = await response.full_response()
                analyst_msg = ChatMessage(
                    role=MessageRole.ASSISTANT,
                    content=f"Analyst: "
                    f"\nHere is the analysis result: {msg_content}"
                    "\nUse it for other steps or response the content to the user.",
                )
                self.memory.put(analyst_msg)
                return InputEvent(input=self.memory.get())
            else:
                tool_calls = response.tool_calls
                self.memory.put(response.tool_call_message)

        # Call tools
        tool_call_outputs = await call_tools(
            ctx=ctx,
            agent_name="Analyst",
            tools=[self.code_interpreter_tool],
            tool_calls=tool_calls,  # type: ignore
        )
        for tool_call_output in tool_call_outputs:
            self.memory.put(
                ChatMessage(
                    role=MessageRole.TOOL,
                    content=tool_call_output.tool_output.content,
                    additional_kwargs={
                        "name": tool_call_output.tool_output.tool_name,
                        "tool_call_id": tool_call_output.tool_call_id,
                    },
                )
            )

        # Fallback to the input with the latest chat history
        return InputEvent(input=self.memory.get())

    @step()
    async def report(self, ctx: Context, ev: ReportEvent) -> InputEvent:
        """
        Generate a report based on the analysis result.
        """
        ctx.write_event_to_stream(
            AgentRunEvent(
                name="Reporter",
                msg="Starting report generation",
            )
        )
        tool_calls = ev.input
        tool_call_outputs = await call_tools(
            ctx=ctx,
            agent_name="Reporter",
            tools=[self.document_generator_tool],
            tool_calls=tool_calls,
        )
        for tool_call_output in tool_call_outputs:
            self.memory.put(
                ChatMessage(
                    role=MessageRole.TOOL,
                    content=tool_call_output.tool_output.content,
                    additional_kwargs={
                        "name": tool_call_output.tool_output.tool_name,
                        "tool_call_id": tool_call_output.tool_call_id,
                    },
                )
            )
        # After the tool calls, fallback to the input with the latest chat history
        return InputEvent(input=self.memory.get())
